src/os/bluestore/BlueFS.cc

```diff
@@ -3022,3 +3161,268 @@ bool BlueFS::wal_is_rotational()
+        if (file) {
+          file_move_start(file, BDEV_DB);
+        }
+      }
+    }
+  } else if (frac < no_slow_on_db) {
+    if (size_on_slow > 0) {
+      file = find_move_candidate(BDEV_SLOW, BDEV_DB);
+      if (file) {
+        file_move_start(file, BDEV_DB);
+      }
+    }
+  } else {
+    file = find_move_candidate(BDEV_DB, BDEV_SLOW);
+    if (file) {
+      file_move_start(file, BDEV_SLOW);
+    }
+  }
+}
+
+BlueFS::FileRef BlueFS::find_move_candidate(uint8_t is_on, uint8_t wants_on)
+{
+  mempool::bluefs::unordered_map<uint64_t,FileRef>::iterator it;
+  it = file_map.find(move_last_ino);
+  if (it != file_map.end())
+    it++;
+  if (it == file_map.end())
+    it = file_map.begin();
+  if (it == file_map.end()) {
+    dout(1) << __func__ << " no files" << dendl;
+    return FileRef{};
+  }
+  
+  for (uint64_t cnt = 0; cnt < 1000; cnt++) {
+    bluefs_fnode_t& fnode = it->second->fnode;
+    if (fnode.prefer_bdev == wants_on &&
+        fnode.extents.size() != 0 &&
+        fnode.extents[0].bdev == is_on) {
+      move_last_ino = it->first;
+      return it->second;
+    }
+    it++;
+    if (it == file_map.end())
+      it = file_map.begin();
+    if(it->first == move_last_ino)
+      break;
+  };
+  move_last_ino = it->first;
+  dout(30) << __func__ << " (from bdev " << is_on << " to bdev "
+      << wants_on << ") candidate not found" << dendl;
+  return FileRef{};
+}
+
+/*
+ * 1. Create target extents, allocate on target device
+ * 2. Copy from f->extents to extents
+ * 3. If f is still existing swap f->extents and extents
+ * 4. Free extents
+ */
+
+BlueFS::AioMoveFileContext::AioMoveFileContext(BlueFS* bfs)
+  : bluefs(bfs),
+    ioc(bfs->cct, this, true)
+{
+}
+
+
+BlueFS::AioMoveFileContext* BlueFS::file_move_start(FileRef f, unsigned target_device)
+{
+  dout(10) << __func__ << " on " << f->fnode << " to bdev " << target_device << dendl;
+  PExtentVector temp_extents;
+  ceph_assert(target_device < alloc.size());
+  ceph_assert(alloc[target_device]);
+  uint64_t need = f->fnode.size;
+  int r = _allocate_without_fallback(target_device,
+                             need,
+                             &temp_extents);
+  if (r != 0) {
+    //skip copying file
+    alloc[target_device]->release(temp_extents);
+    return nullptr;
+  }
+
+  AioMoveFileContext* c = new AioMoveFileContext(this);
+  c->source = f;
+  c->source_it = c->source->fnode.extents.begin();
+  c->source_offset = 0;
+  for (auto& i:temp_extents) {
+    c->target.emplace_back(target_device, i.offset, i.length);
+  }
+  c->target_it = c->target.begin();
+  c->target_offset = 0;
+  //bootup copying
+  move_context = c;
+  aio_read_chunk(c);
+  return c;
+}
+
+void BlueFS::file_move_finished(AioMoveFileContext* c)
+{
+  dout(10) << __func__ << " on " << c->source->fnode.ino << " result=" << c->result << dendl;
+  ceph_assert(move_context == c);
+  bool needsync = false;
+  if (c->result) {
+    lock.lock();
+    if (c->source->num_reading.try_lock()) {
```

Comment:
I don't completely understand what's the handling here. What happens if somebody is reading from the file? You silently undo the move? What happens if file has been changed during the move? You don't hold exclusive lock over it while moving,do you?

Reply:
@ifed01 You exactly touched most critical part.
Skipping for moment the fact that check if file has changed is missing,
do you think that copying file in background and on final step either committing it, or dropping it is valid?

Reply:
@aclamk - I don't see any unsolvable issues with this approach, looks viable...